File: app/src/main/java/com/banglalink/toffee/model/Channel.java
package com.banglalink.toffee.model;

import android.content.Context;
import android.net.Uri;
import android.os.Bundle;

import com.banglalink.toffee.data.storage.Preference;
import com.banglalink.toffee.ui.player.Samples;
import com.banglalink.toffee.util.Utils;

import java.net.MalformedURLException;
import java.net.URL;

import static com.google.android.exoplayer2.C.TYPE_HLS;

/**
 * Created by shantanu on 8/31/16.
 */

public class Channel extends Samples.Sample {

    public Channel(String name, String uri, int type, String imageUrl) {
        super(name, uri, type);
        this.imageUrl = imageUrl;
    }

    public Channel(String name, String uri) {
        super(name, uri, TYPE_HLS);
        this.imageUrl = "";
    }

    public static Channel createChannel(ChannelInfo channelInfo){
        return new Channel(channelInfo.program_name,channelInfo.hlsLinks.get(0).hls_url_mobile);
    }

    public Channel(String name, String contentId, String provider, String uri, int type, String imageUrl) {
        super(name,contentId,provider,uri,type);
        this.imageUrl = imageUrl;
    }

    public final String imageUrl;

    public Bundle getBundle(){
        Bundle bundle = new Bundle();
        bundle.putString("name",name);
        bundle.putString("contentid",contentId);
        bundle.putString("provider",provider);
        bundle.putString("uri",uri);
        bundle.putInt("type",type);
        bundle.putString("imageurl",imageUrl);
        return bundle;
    }

    public static Channel create(Bundle bundle){
        Channel channel = new Channel(bundle.getString("name"),
                bundle.getString("contentid"),
                bundle.getString("provider"),
                bundle.getString("uri"),
                bundle.getInt("type"),
                bundle.getString("imageurl"));
        return channel;
    }

    public String getContentUri(Context context){
        Uri contentUri = null;
        boolean isWifiConnected = Utils.checkWifiOnAndConnected(context);
        if(!isWifiConnected && Preference.Companion.getInstance().watchOnlyWifi()){
            return null;
        }
        String text = "";
        if (isWifiConnected) {
            if (Preference.Companion.getInstance().getWifiProfileStatus() == 6) {
                text = "/auto";
            } else {
                text = "/" + Preference.Companion.getInstance().getWifiProfileStatus();
            }
        } else {
            text = "/" + Preference.Companion.getInstance().getCellularProfileStatus();
        }

        if(Preference.Companion.getInstance().shouldOverrideHlsUrl()){
            try {
                String path = new URL(uri).getPath();
                uri = Preference.Companion.getInstance().getHlsOverrideUrl()+path;
            } catch (MalformedURLException e) {
                e.printStackTrace();
            }
        }
        if(uri.endsWith("/")){
            contentUri = Uri.parse(uri + Preference.Companion.getInstance().getSessionToken()+ text);
        }
        else {
            contentUri = Uri.parse(uri + "/" + Preference.Companion.getInstance().getSessionToken() + text);
        }
        return contentUri.toString();
    }
}

File: app/src/main/java/com/banglalink/toffee/model/Channel.kt
package com.banglalink.toffee.model

import android.content.Context
import android.net.Uri
import android.os.Bundle
import android.text.TextUtils
import com.banglalink.toffee.data.storage.Preference
import com.banglalink.toffee.ui.player.Samples
import com.banglalink.toffee.util.Utils
import com.google.android.exoplayer2.C
import java.net.MalformedURLException
import java.net.URL
import java.util.*

/**
 * Created by shantanu on 8/31/16.
 */
class Channel(
    private val name: String?,
    contentId: String? = name?.toLowerCase(Locale.US)?.replace("\\s".toRegex(), ""),
    provider: String? = "",
    uri: String? = null,
    type: Int = C.TYPE_HLS,
    private var imageUrl: String? = null
): Samples.Sample(name, contentId, provider, uri, type)
{

    val bundle: Bundle =  Bundle().apply {
        putString("name", name)
        putString("contentid", contentId)
        putString("provider", provider)
        putString("uri", uri)
        putInt("type", type)
        putString("imageurl", imageUrl)
    }

    fun getContentUri(context: Context, pref: Preference): String? {
        val isWifiConnected = Utils.checkWifiOnAndConnected(context)
        if (!isWifiConnected && pref.watchOnlyWifi()) {
            return null
        }
        val text = if (isWifiConnected) {
            if (pref.wifiProfileStatus == 6) {
                "/auto"
            } else {
                "/" + pref.wifiProfileStatus
            }
        } else {
            "/" + pref.cellularProfileStatus
        }
        if (pref.shouldOverrideHlsUrl()) {
            try {
                val url = URL(uri)
                var path = url.path
                if (!TextUtils.isEmpty(url.query)) {
                    path = path + "?" + url.query
                }
                uri = pref.getHlsOverrideUrl() + path
            } catch (e: MalformedURLException) {
                e.printStackTrace()
            }
        }
        val contentUri = if (uri.endsWith("/")) {
            Uri.parse(uri + pref.sessionToken + text)
        } else {
            Uri.parse(uri + "/" + pref.sessionToken + text)
        }
        return contentUri.toString()
    }

    companion object {
        @JvmStatic
        fun createChannel(channelInfo: ChannelInfo): Channel {
            return Channel(channelInfo.program_name, uri = channelInfo.hlsLinks!![0].hls_url_mobile)
        }

        fun create(bundle: Bundle): Channel {
            return Channel(
                bundle.getString("name"),
                bundle.getString("contentid"),
                bundle.getString("provider"),
                bundle.getString("uri"),
                bundle.getInt("type"),
                bundle.getString("imageurl")
            )
        }
    }
}